### DISCUSSION TITLE
How to host an online version of BIPES

### DISCUSSION THREAD

**Discussion by gpeyton** [BIPES/BIPES#103] [Q&A] [answered]
Thanks for the great work on BIPES!

How do you go about hosting a forked version of BIPES online? (Including features such as MQTT).

**Comment by gastmaier** [comment, marked as answer]
Well, that is a trick question! BIPES 2 source code doesn't include the "MQTT server" (it's mosquitto + PHP + PostgreSQL) because it was done outside the repository and never officially included.
I will take a look around if I can provide a zip with these files for now...

The repository files you just have to make it accessible  by an [Apache](https://ubuntu.com/tutorials/install-and-configure-apache) server or similar.

However, big news... coming soon! The next release will run inside a Python virtual environment with the server running in Flask, with development mode, self setup MQTT and everything! I'm just taking my time for now in a private repository, so I don't publish with regretful design decisions that I couldn't change after making it public! 😅

**Comment by rafaelaroca** [comment]
Hello @gpeyton, here is a complete tarball with all sources included (as it is on the server):

https://bipes.net.br/beta2/bipes_full_11_02_2022.tar.gz

If you copy this onto another server public HTML directory, you will have most BIPES features working on another host. As @gastmaier said, you will also need the MQTT server running on the same machine and adjust easymqtt files (PHP scripts) to access the correct database.

There is another tricky part: code sharing by URL uses google app engine, so you must deploy the code sharing to a google app engine account (the file is ./appengine/bipes-271213/storage.js on the tarball we sent).

hope it helps a bit more

[]s Rafael

**Reply by gpeyton** [reply]
@gastmaier @rafaelaroca - thanks for the feedback, and looking forward to the next release!

**Reply by gastmaier** [reply]
Hey @gpeyton the code is public, to try it out do:
```
git clone -b third git@github.com:BIPES/BIPES.git
cd BIPES && make
```
then follow the prompt.
To print a brief description of other commands, do  `make help`.

Be aware that this is an alpha, many blocks and devices haven't been ported from BIPES2 yet and is missing some features.

Also, I'm having issues reaching the self-hosted mosquitto broker, but I think the problem is my pc/router.

PS. Right click stuff, BIPES3 uses a custom context menu to provide more features without making the UI bloated.

PS2. Here is a print from yesterday, where I shared the same device between two tabs to acquire the [PID](https://github.com/gastmaier/micropython-simple-pid) response of a thermal plant.

![print](https://user-images.githubusercontent.com/2892061/156174341-93e4a7cc-2e22-4b93-b20e-59ba84a9b37d.png)

Cool new features:
 - Share devices connection between tabs/windows.
 - Work in the same project or multiple projects in multiple tabs/windows (every change is dispatched, keeping projects in sync).
 - Full support to device's directories, making easy to manage libraries.
 - ctrl+alt+c and ctrl+alt+v, copy and paste code in the terminal, without messing indentation anymore.
 - Both MQTT and Serial (localStorage/serverless) real time plotting, storing all incoming data as a topic.
 - Multiple programming languages support.
 - Supports deploying a minimized version with a single command (`make release`***), that doesn't require a server.
 - Development server written in python:
   - Easy to orchestrate anything (OpenCV? Custom firmware? Virtual machines?).
   - Templates, so files won't have 10k lines anymore.
  - Aims to be 100% self-hosted. 
  - Written in vanilla JavaScript, (no react.js, vue.js, bootstrap, bloat.js...).

**Reply by gpeyton** [reply]
Excellent work, and thanks for the update! 
- How will multiple languages work if you're planning to work with MicroPython/CircuitPython? I guess you will need to restrict to Python in this case?
- Is the minimized version like the old "offline" build?

**Reply by gastmaier** [reply]
1. BIPES as is will restrict the blocks to pythonic languages (MicroPython, CircuitPython and Snek), I personally think anything more than that would be too much work with little gain.
2. Yes. Actually, the  `make release` minimizes and generates a zip (like the old  "offline" build) and the `make deploy` minimizes and deploys to a web server (eg. apache2).
 
Context on 2:
Consider a teacher wanting to give robotics classes, and he has available ESP32, ST Nucleo and some sensors, It makes sense to, with the development mode, remove the other devices, make some adjustments here and there and then deploy to the schools' server. If running in a  local server is not an option, he can still give his students the zip, and they can run directly from the `ide.html` file

More on 1:
However, others can easily make their own fork supporting other languages,  [static/page/blocks/pythonic](https://github.com/BIPES/BIPES/tree/third/static/page/blocks/pythonic) contains the files to generate code for MicroPython, CircuitPython and Snek. Someone could easily do a static/page/blocks/cpp to support C++. To interface a REPL (fetch files, make directories), there is [static/base/rosetta.js](https://github.com/BIPES/BIPES/blob/third/static/base/rosetta.js) to interface each language.

**Comment by julienrat** [comment]
Works like a charm !
